### ISSUE TITLE
Encountered an issue while generating pre-signed URL while getting image from s3 bucket

### ISSUE THREAD

**Issue by marenero1974** [aws-amplify/amplify-android#1267]
Hi all, I have a problem while getting an image from my S3 bucket:

Here is my code: 

```java
ImageView image = mView.findViewById(R.id.image);
Amplify.Storage.getUrl(imagePath,
    result -> {
        URL url = result.getUrl();
        Glide.with(ctx)
            .load(url)
            .apply(new RequestOptions()
                .diskCacheStrategy(DiskCacheStrategy.NONE)
                .skipMemoryCache(true))
                .into(image);
            Log.i("Amplify---------", "url ok----");
        },
        error -> Log.i("Amplify---------", "error while retrieving url: " + error.getMessage())
);

````

In my logs i get:
```console
error while retrieving url: Encountered an issue while generating pre-signed URL
```
and the image won't show.

Please can anyone provide a solution or fix?

Thanks

**Comment by drochetti** [comment]
Hi @marenero1974, can you provide more details of your setup?

- Amplify version
- Have you setup the Auth plugin as well? The [Provision backend storage](https://docs.amplify.aws/lib/storage/getting-started/q/platform/android#provision-backend-storage) guide
- Can guest users access the files on your bucket (this is a setup option in `amplify add storage`)?
- Can you log the entire error and share that with us, instead of jut the message? `Log.e(error)`

**Comment by marenero1974** [comment]
@drochetti 
Amplify version: 1.17.2, yes i setup Auth like this:

````
            Amplify.addPlugin(new AWSApiPlugin()); // UNCOMMENT this line once backend is deployed
            Amplify.addPlugin(new AWSCognitoAuthPlugin());
            Amplify.addPlugin(new AWSS3StoragePlugin());
            Amplify.addPlugin(new AWSDataStorePlugin());

````

also guest users can access files, i have added with amplify add storage
these are few logs of my application:

````

2021-04-06 07:17:02.111 31211-31430/com.my.app I/amplify:aws-datastore: Orchestrator lock released.
2021-04-06 07:17:02.220 31211-31444/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->compareAndSwapObject(Ljava/lang/Object;JLjava/lang/Object;Ljava/lang/Object;)Z (greylist, linking, allowed)
2021-04-06 07:17:02.220 31211-31444/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->compareAndSwapObject(Ljava/lang/Object;JLjava/lang/Object;Ljava/lang/Object;)Z (greylist, linking, allowed)
2021-04-06 07:17:02.221 31211-31444/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->putObject(Ljava/lang/Object;JLjava/lang/Object;)V (greylist, linking, allowed)
2021-04-06 07:17:02.221 31211-31444/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->getLong(Ljava/lang/Object;J)J (greylist,core-platform-api, linking, allowed)
2021-04-06 07:17:02.222 31211-31444/com.my.app I/FIAM.Headless: Successfully fetched 0 messages from backend
2021-04-06 07:17:02.231 31211-31457/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->getInt(Ljava/lang/Object;J)I (greylist, linking, allowed)
2021-04-06 07:17:02.231 31211-31457/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->putInt(Ljava/lang/Object;JI)V (greylist, linking, allowed)
2021-04-06 07:17:02.231 31211-31457/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->putLong(Ljava/lang/Object;JJ)V (greylist, linking, allowed)
2021-04-06 07:17:02.235 31211-31444/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->getLong(Ljava/lang/Object;J)J (greylist,core-platform-api, linking, allowed)
2021-04-06 07:17:02.237 31211-31289/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->getInt(Ljava/lang/Object;J)I (greylist, linking, allowed)
2021-04-06 07:17:02.240 31211-31289/com.my.app I/FirebasePerformance: Logging network request trace: https://cognito-identity.eu-west-2.amazonaws.com/ (responseCode: 200, responseTime: 107.2580ms)
2021-04-06 07:17:02.240 31211-31289/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->getInt(Ljava/lang/Object;J)I (greylist, linking, allowed)
2021-04-06 07:17:02.241 31211-31289/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->getLong(Ljava/lang/Object;J)J (greylist,core-platform-api, linking, allowed)
2021-04-06 07:17:02.432 31211-31211/com.my.app I/HwViewRootImpl: removeInvalidNode jank list is null
2021-04-06 07:17:02.461 31211-31367/com.my.app W/libEGL: EGLNativeWindowType 0x7425ec1990 disconnect failed
2021-04-06 07:17:02.466 31211-31457/com.my.app I/Amplify---------: url ok----
2021-04-06 07:17:02.491 31211-31476/com.my.app I/Amplify---------: url ok----
2021-04-06 07:17:02.515 31211-31477/com.my.app I/Amplify---------: error while retrieveing url: Encountered an issue while generating pre-signed URL
2021-04-06 07:17:02.558 31211-31487/com.my.app I/Amplify---------: url ok----
2021-04-06 07:17:02.603 31211-31488/com.my.app I/Amplify---------: error while retrieveing url: Encountered an issue while generating pre-signed URL
2021-04-06 07:17:02.873 31211-31587/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->putObject(Ljava/lang/Object;JLjava/lang/Object;)V (greylist, linking, allowed)
2021-04-06 07:17:02.873 31211-31586/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->putObject(Ljava/lang/Object;JLjava/lang/Object;)V (greylist, linking, allowed)
2021-04-06 07:17:02.879 31211-31289/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->getInt(Ljava/lang/Object;J)I (greylist, linking, allowed)
2021-04-06 07:17:02.879 31211-31289/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->putInt(Ljava/lang/Object;JI)V (greylist, linking, allowed)
2021-04-06 07:17:02.884 31211-31289/com.my.app I/FirebasePerformance: Logging network request trace: https://pubwin2bucketcmsi235143-dev.s3.eu-west-2.amazonaws.com/public/Profile%20Images/iLO9DjDFslRD6RA5bK9EmtdYk0K2.jpg (responseCode: 200, responseTime: 294.3300ms)
2021-04-06 07:17:02.884 31211-31289/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->getInt(Ljava/lang/Object;J)I (greylist, linking, allowed)
2021-04-06 07:17:02.884 31211-31289/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->getLong(Ljava/lang/Object;J)J (greylist,core-platform-api, linking, allowed)
2021-04-06 07:17:02.892 31211-31289/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->getInt(Ljava/lang/Object;J)I (greylist, linking, allowed)
2021-04-06 07:17:02.894 31211-31289/com.my.app I/FirebasePerformance: Logging network request trace: https://pubwin2bucketcmsi235143-dev.s3.eu-west-2.amazonaws.com/public/Profile%20Images/iLO9DjDFslRD6RA5bK9EmtdYk0K2.jpg (responseCode: 200, responseTime: 297.6590ms)
2021-04-06 07:17:02.894 31211-31289/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->getInt(Ljava/lang/Object;J)I (greylist, linking, allowed)
2021-04-06 07:17:02.894 31211-31289/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->getLong(Ljava/lang/Object;J)J (greylist,core-platform-api, linking, allowed)
2021-04-06 07:17:02.898 31211-31289/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->getInt(Ljava/lang/Object;J)I (greylist, linking, allowed)
2021-04-06 07:17:02.903 31211-31289/com.my.app I/FirebasePerformance: Logging network request trace: https://pubwin2bucketcmsi235143-dev.s3.eu-west-2.amazonaws.com/public/Profile%20Images/oZWrgbXyPyOug88ckw8ELGWucuU2.jpg (responseCode: 200, responseTime: 268.1040ms)
2021-04-06 07:17:02.904 31211-31289/com.my.app W/my.app: Accessing hidden method Lsun/misc/Unsafe;->getInt(Ljava/lang/Object;J)I (greylist, linking, allowed)


````

**Comment by drochetti** [comment]
Hey @marenero1974 thanks for the extra info. Unfortunately the error message alone is not telling me enough. Would you mind sharing the entire error stack trace?

So instead of `Log.i("Amplify---------", "error while retrieving url: " + error.getMessage())` do `Log.e(error)` and you should get the stack trace.

**Comment by marenero1974** [comment]
Thanks @drochetti : the problem was:  

`error while retrieveing url: java.lang.IllegalArgumentException: You must call this method on the main thread`

now i changed with:

```
ImageView image = mView.findViewById(R.id.image);
Amplify.Storage.getUrl(imagePath,
    result -> {
        URL url = result.getUrl();
        runOnUiThread(() -> Glide.with(ctx)
            .load(url)
            .apply(new RequestOptions()
                .diskCacheStrategy(DiskCacheStrategy.NONE)
                .skipMemoryCache(true))
                .into(image));
            Log.i("Amplify---------", "url ok----");
        },
        error -> Log.e("Amplify---------", "error while retrieveing url: " + error.getCause().toString()));
```

but now i have another issue: there are no exceptions but i display my images in a RecyclerView only once: i put this code in a method of a custom adapter i made, and the adapter is added in a recyclerView in my activity like this:

```
        LinkedList<Table> list = new LinkedList<>();
        Adapter adapter = new Adapter(list, MainActivity.this);
        Amplify.DataStore.query(
            Table.class,
            items -> {
                while (items.hasNext()) {
                    Table table = items.next();
                    list.add(table);

                    Log.i("Amplify", "Table Id " + table.getId());
                }
            },
            failure -> Log.e("Amplify", "Could not query DataStore", failure)
        );

        adapter.setMessages(list);
        recyclerView.setAdapter(adapter);
        adapter.notifyDataSetChanged();	
```

Where am i wrong?

**Comment by marenero1974** [comment]
I solved the problem, it was mine problem not related to amplify, thanks @drochetti
